### ISSUE TITLE
Finish voicecom

### ISSUE THREAD

**Issue by fredi-68** [fredi-68/Ram#6]
Oh boy.

**Comment by fredi-68** [comment]
Right, so I'm just gonna leave some thoughts here while I work on this.

Done some more extensive testing and here's the summary.
- Reading from the UDP voice socket still crashes the event loop upon leaving the channel on Windows. This does NOT happen on Linux so I suspect a bug in the Windows implementation of the Selector loop in asyncio to be the cause.
- I can still not get the UDP socket to receive any data AT ALL. Somehow, during IP discovery, discord.py manages to get a packet response from the server, then it's complete silence. I'd blame it on NAT wasn't it for the fact that the voice connection actually is established, even if in one direction only. I feel like we're missing an additional packet of sorts to signal the server that we want to start receiving voice data but what this packet looks like and whether it is sent through the WS or UDP I do not know.

**Comment by fredi-68** [comment]
Just in case this is important later, I figured out the reason that the voice socket wasn't receiving any data: Apparently you are required to send at least five frames of audio data on the socket before the server recognizes you as a participant. Before this, no data will be sent by the server, which was happening in testing, causing a lot of confusion.
In light of the changes to discord.py I decided to completely rewrite the voicecom module from scratch and make it a package while I was at it. I have voice receive working completely now and audio signals can be extracted in real time, and on a per user basis. Still missing are a convenient interface to a channel mixdown, yielding all speakers audio while providing best effort timing. This will most likely require extensive buffering so making it a separate interface is my best idea right now. Speech recognition APIs using the existing interface are also implemented, though I am struggling to find a service provider I can actually use that won't cause any issues. The only offline, self hosted library I have tried doesn't have very high accuracy since it uses a limited HMM, as opposed to modern deep learning architectures like google cloud speech. Since I would also like use google for text to speech at a later step, I may have to figure out a way to make this work after all.
Other things that are left is making an interface for commands and the conversation simulator that is accessible through voice. There are two options here: Creating a fake Message instance, which is convenient but may lead to issues, or modifying the ResponseManager to support audio as an input source.
